File: program/src/objects/Engine.cpp
/*
 * Steer.h
 *
 *  Created on: Jun 8, 2016
 *      Author: gijs
 */

#include "Engine.h"
#include "Pwm.h"
#include <syslog.h>

static PWM pwm;
static int dutyMax= 2000000;
static int dutyMiddle = 1400000;
static int dutyMin= 800000;

#define STEP 50000

EngineEvent::EngineEvent(int throttle) {
	this->throttle = throttle;
}

EngineEvent::~EngineEvent() {

}

int EngineEvent::getThrottle() {
	return this->throttle;
}

Engine::Engine(EventHandler *eventHandler) {
	syslog(LOG_INFO, "%s", "Setting up the engine.");
	int pwmNr = 0;
	int period = 20000000;
	pwm = PWM(pwmNr);
	if (pwm.setPeriod(period) > 0) {
		syslog(LOG_ERR, "Could not set pwm period: %s", pwm.getName());
	}
	if (pwm.setDuty(dutyMiddle) > 0) {
		syslog(LOG_ERR, "Could not set pwm duty: %s", pwm.getName());
	}
	if (pwm.setPolarity(0) > 0) {
		syslog(LOG_ERR, "Could not set pwm polarity: %s", pwm.getName());
	}
	if (pwm.start() > 0) {
		syslog(LOG_ERR, "Could not start pwm port: %i", pwmNr);
	}
	throttle = dutyMiddle;
	this->eventHandler = eventHandler;
	syslog(LOG_INFO, "%s", "Here is the engine.");
}

Engine* Engine::getInstance(EventHandler *eventHandler) {
	static Engine obj(eventHandler);
	return &obj;
}

void Engine::setThrottle(int t) {
	if (pwm.setDuty(t) > 0) {
		syslog(LOG_INFO,"%s","Could not set throttle.");
	} else {
		syslog(LOG_DEBUG,"Throttle set to: %i", throttle);
		throttle = t;
		Event *event = new EngineEvent(throttle);
		this->eventHandler->handleEvent(event);
	}
}

int Engine::getThrottle() {
	return throttle;
}

void Engine::speedUp() {
	if (throttle + STEP <= dutyMax)
		setThrottle(throttle + STEP);
}

void Engine::slowDown() {
	if (throttle - STEP >= dutyMin)
		setThrottle(throttle - STEP);
}

File: src/main/java/nl/carmageddon/domain/Engine.java
package nl.carmageddon.domain;

import com.google.inject.Inject;
import com.google.inject.name.Named;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Singleton;
import java.util.Observable;

/**
 * @author Gijs Sijpesteijn
 */
@Singleton
public class Engine extends Observable {
    private static final Logger log = LoggerFactory.getLogger(Engine.class);
//    private static int dutyMin= 800000;
//    private static int dutyMax= 2000000;
    private static int dutyMiddle = 1400000;
    private static int MAX_THROTTLE = 120;
    private static int SPEED_STEP = 5000;

    private Pwm pwm;
    private int throttle;
    private int throttleLimit = 20;
    private boolean connected;

    @Inject
    public Engine(@Named("PWM22") Pwm pwm) {
        this.pwm = pwm;
        if (this.pwm != null) {
            this.pwm.setPeriod(20000000);
            this.pwm.setPolarity(0);
            this.connected = true;
            this.setThrottle(0);
            this.connected = false;
            this.pwm.start();
        }
    }

    public int getThrottle() {
        return (throttle - dutyMiddle)/SPEED_STEP;
    }

    // TODO Check of stap niet te groot is. soms stopt de bb als je abrupt van max naar min gaat.
    // Misschien wordt er dan te veel vermogen gevraagd voor de motor en is er niet genoeg voor de bb.
    public void setThrottle(int throttle) {
        if (!connected) {
            log.error("I won't set the throttle, because no clients are connected.");
            return;
        }
        if (throttle < -throttleLimit) {
            throttle = -throttleLimit;
        }
        if (throttle > throttleLimit) {
            throttle = throttleLimit;
        }
        log.debug("Setting throttle: " + throttle);
        this.throttle = dutyMiddle + ( throttle * SPEED_STEP);
        this.pwm.setDuty(this.throttle);
        setChanged();
        notifyObservers();
    }

    public void speedUp() {
        int speedUp = this.getThrottle() - 1;
        setThrottle(speedUp);
    }

    public void slowDown() {
        int slowDown = this.getThrottle() + 1;
        setThrottle(slowDown);
    }

    public void setConnected(boolean connected) {
        if(!connected) {
            setThrottle(0);
        }
        this.connected = connected;
    }

    public int getThrottleLimit() {
        return throttleLimit;
    }

    public void setThrottleLimit(int throttleLimit) {
        if (throttleLimit <= MAX_THROTTLE)
            this.throttleLimit = throttleLimit;
    }
}
